nixos/modules/services/web-apps/keycloak.nix

```diff
@@ -0,0 +1,441 @@
+        export JBOSS_BASE_DIR="$(pwd -P)";
+        export JBOSS_MODULEPATH="${cfg.package}/modules";
+        export JBOSS_LOG_DIR="$JBOSS_BASE_DIR/log";
+
+        cp -r ${cfg.package}/standalone/configuration .
+        chmod -R u+rwX ./configuration
+
+        mkdir -p {deployments,ssl}
+
+        "${cfg.package}/bin/standalone.sh"&
+
+        attempt=1
+        max_attempts=20
+        while ! ${cfg.package}/bin/jboss-cli.sh --connect ':read-attribute(name=server-state)'; do
+            echo "Keycloak not fully started yet, waiting 3 sec.."
+            if [[ "$attempt" -eq "$max_attempts" ]]; then
+                exit 1
+            fi
+            (( attempt++ ))
+            sleep 3
+        done
+
+        attempt=1
+        max_attempts=5
+        while ! ${cfg.package}/bin/jboss-cli.sh --connect --file=${jbossCliScript} --echo-command; do
+            if [[ "$attempt" == "$max_attempts" ]]; then
+                exit 1
+            fi
+            (( attempt++ ))
+            echo "Configuration application failed, retrying.. ($attempt/5)" >&2
+            sleep 3
+        done
+
+        cp configuration/standalone.xml $out
+      '';
+    in
+      lib.mkIf cfg.enable {
+
+        environment.systemPackages = [ cfg.package ];
+
+        systemd.services.keycloakDatabaseInit = lib.mkIf databaseActuallyCreateLocally {
+          after = [ "postgresql.service" ];
+          before = [ "keycloak.service" ];
+          bindsTo = [ "postgresql.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            RemainAfterExit = true;
+            User = "postgres";
+            Group = "postgres";
+          };
+          script = ''
+            set -eu
+
+            if ! [[ -e ${cfg.databasePasswordFile} ]]; then
+                echo "Database password file ${cfg.databasePasswordFile} not found! Aborting." >&2
+                exit 1
+            fi
+
+            PSQL=${config.services.postgresql.package}/bin/psql
+
+            db_password="$(<'${cfg.databasePasswordFile}')"
+
+            $PSQL -tAc "SELECT 1 FROM pg_roles WHERE rolname='${cfg.databaseUsername}'" | grep -q 1 || $PSQL -tAc "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN PASSWORD '$db_password' CREATEDB"
+
+            $PSQL -tAc "SELECT 1 FROM pg_database WHERE datname = 'keycloak'" | grep -q 1 || $PSQL -tAc 'CREATE DATABASE "keycloak" OWNER "${cfg.databaseUsername}"'
+            current_owner=$($PSQL -tAc "SELECT pg_catalog.pg_get_userbyid(datdba) FROM pg_catalog.pg_database WHERE datname = 'keycloak'")
+            if [[ "$current_owner" != "${cfg.databaseUsername}" ]]; then
+                $PSQL -tAc 'ALTER DATABASE "keycloak" OWNER TO "${cfg.databaseUsername}"'
+                if [[ -e "${config.services.postgresql.dataDir}/.reassigning_keycloak" ]]; then
+                    echo "Reassigning ownership of database keycloak to user ${cfg.databaseUsername} failed on last boot. Failing..." >&2
```

Comment:
This is scary to me. I won't make a big deal about it, but I *personally* don't think we should reassign sole ownership databases this way.

Reply:
I've removed this this and now force the local db username to be `keycloak` - I figure there isn't really a need for it to be configurable when a local db is provisioned automatically.